### ISSUE TITLE
Correct usage of rospy.spin()

### ISSUE THREAD

**Issue by Kshitij1K** [Kshitij1K/personal-issue-log#4]
Finally understood when to use rospy.spin()

TLDR: You don't need it. The code will just work fine without it, as long as an infinite loop keeps running in main().

Common mistake I do every time I switch to python after a long time:

```
rospy.init_node('some_node')
r = rospy.Rate(10)

while not rospy.is_shutdown():
    # Do something, may be publish some stuff
    rospy.spin()
    r.sleep()
```
DON'T DO THIS, this will block the while loop at rospy.spin(). rospy.spin() should ONLY be placed at the end of the code, since it is basically an infinite loop that keeps looping till CTRL-C is not pressed.

The name is a misnomer, it is NOT like the C++ ros::spin(), which  ensures that callback functions are continuously called. 
If you have the following code in your main function:
```
rospy.init_node('some_node')
r = rospy.Rate(10)

while not rospy.is_shutdown():
    # Do something, may be publish some stuff
    r.sleep()
```

you don't need rospy.spin(). The callbacks run in a separate thread, and are automatically called if a message is received on a topic.

rospy.spin() should only be used if you want to do the following:
```
rospy.init_node('some_node')

while not rospy.is_shutdown():
    # You have nothing to do here
```
rospy.spin is a nice condensed way to write the above code. It just ensures that the program keeps running.
